compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp

```diff
@@ -363,6 +363,90 @@ CompilationInfoAttr::verify(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// iree_codegen.workgroup_mapping
+//===----------------------------------------------------------------------===//
+
+WorkgroupMappingAttr WorkgroupMappingAttr::get(MLIRContext *context,
+                                               WorkgroupId id) {
+  return WorkgroupMappingAttr::get(context, id, /*delinearizedDim=*/0);
+}
+
+LogicalResult
+WorkgroupMappingAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+                             WorkgroupId id, int64_t delinearizedDim) {
+  if (delinearizedDim > 0 && id != WorkgroupId::IdZ) {
+    return emitError() << "illegal to use `delinearizationDim` for "
+                       << stringifyWorkgroupId(id);
```

Comment:
why is delinearization only legal for `z`?

Reply:
IMO, the x, y, z itself is a dilinearization of ids. So I dont see much value in have delinearization of x and y separately. You need delinearization on z to be able to map "logical 4D" to these three dimensions. Also makes the resolution logic easier, cause you handle x and y in a particular way and any delinearization you need to do on top of that just map that to Z. I have only seen this usage in practice.

Reply:
ah I see, that makes sense because the result of workgroup tiling sets the sizes. I guess the same thing doesn't work the same way on GPU because there might be multiple different logical bases (i.e. multiple different scf.forall ops), but maybe it would be a good idea to try to set the actual 3D workgroup size later in the pipeline and only set the flat 1D workgroup size in KernelConfig.

Reply:
maybe.... I dont think I fully grok all of that, but if you have an example handy it will help me anchor what you are saying.